### ISSUE TITLE
Type-lambda syntax

### ISSUE THREAD

**Issue by puffnfresh** [typelevel/scala#6]
We should introduce simplified type-lambda syntax, for example:

```
[α] = Either[A, α]
```

Which would desugar to:

```
({type λ[α] = Either[A, α]})#λ
```

**Comment by jdegoes** [comment]
In Dotty, wouldn't this be `Either[A, _]`, which desugars to `({type λ[α] = Either[A, α]})#λ`?

**Comment by YoEight** [comment]
Yes please :+1: 

@jdegoes

It seems that it can overlap with how you can express existential type, am I wrong ? (I mean in Scala)

**Comment by vendethiel** [comment]
@YoEight I don't think it can conflict: they don't appear in the same place (sgnature VS expression)

**Comment by tixxit** [comment]
@jdegoes You can use something like that for some cases, but the [x] = Either[A, x] approach is still required in some cases (eg. where x is repeated, for instance, or perhaps nested).

**Comment by tixxit** [comment]
Of course, @non knows all about this, having implemented the compiler plugin. Looking forward to what comes out of this, since now we aren't required to have place holder types and whatnot.

**Comment by jdegoes** [comment]
Sure, we should just try to be forward-compatible with future versions of Scala as laid out in the roadmap (e.g. _both_ `Either[A, _]` and `[x] = Either[A, x]`), at least where it makes sense to do so.

**Comment by tixxit** [comment]
Oh yeah, that I agree with. The kind projector plugin supports both too (but uses ? instead of _)

**Comment by propensive** [comment]
@jdegoes Dotty isn't on our radar at all. It won't be binary compatible, source compatible, or merge compatible, so it's kind of a lost cause.

**Comment by jdegoes** [comment]
@propensive Sorry, not Dotty, but the currently hypothetical "Typesafe Scala 2018".

**Comment by non** [comment]
I don't see how we can use `List[_]` to mean `({type L[X] = List[X]})#L` given that it already means `List[X] forSome {type X}`. Kind Projector uses `?` for this reason. Am I missing something?

**Comment by jdegoes** [comment]
@non A type annotation for a value is always an existential type, while a type declaration or type parameter is always a type lambda. 

e.g.

``` scala
type EitherException[_] = Either[Exception, _] // type lambda
val foo: Either[Exception, _] // existential
bar[Either[Exception, _]] // type lambda
```

**Comment by puffnfresh** [comment]
@jdegoes doesn't that conflict with ascribing existential types?

``` scala
List[List[_]](List(1, 2, 3), List("Hello", "world")).map(_.length)
```

**Comment by jdegoes** [comment]
I think you can disambiguate that by generalizing in the following fashion: where a type of kind `*` is expected, the interpretation is existential; and where a type of a higher kind is expected, the interpretation is a type lambda.

I don't know how the Typesafe guys are planning to do this but I've read from multiple sources that '_' will be used for both existential use cases as well as type lambda use cases (or at least, the subset of type lambda cases that this approach can handle).

**Comment by non** [comment]
My plan for this is to support an explicit type lambda syntax, e.g. something like:
- `[A] Either[A, Int]`
- `[A] => Either[A, Int]`
- `[A] = Either[A, Int]`

Once that is working I'm happy to start looking into supporting some kind of placeholder syntax. Part of me thinks the overhead of internalizing more complicated rules for the meaning of `_` is not worth the trouble, but maybe I'm too pessimistic.

I'll definitely ask some folks at Typesafe about their plans here.

**Comment by tixxit** [comment]
@non Yeah, I agree. Certainly worth just getting the type lambda syntax working, as that would make me super happy. I'll vote for the 2nd syntax (`[A] => Either[A, Int]`).

**Comment by jdegoes** [comment]
+1 on the 2nd syntax for consistency with value lambdas.

**Comment by non** [comment]
I think I like the 2nd best too, but these are three options that had been thrown around so I didn't want to just assume that my preference should carry the day.

(Also, I think Scala may already use the first internally in some cases.)

**Comment by puffnfresh** [comment]
@non I definitely prefer the first, since that's what scalac gives in error messages. If we can toString it, we should be able to parse it!

**Comment by propensive** [comment]
I much prefer the second, and suggest changing the error output format for consistency.

**Comment by paulp** [comment]
`[A] => Either[A, Int]`

The `=>` becomes difficult to read and manipulate in the face of nesting, especially if function types are involved, or much of anything other than the optimistic types we tend to think up when imagining how an idea is going to work rather than how it isn't going to work. In fact this is true of most obvious syntaxes because square brackets nest about as well as dodo birds. You're adding new square brackets and a function arrow even to the simplest partial application. It would be a blizzard of right angular noise. Never has haskell's type application syntax looked better.

Also, it would be a big mistake not to unmake the mistake of confusing existential type arguments with unnamed higher order type parameters. You have to think about more than value and type positions. Underscores also appear in type bounds and in patterns, and you don't always have an expected kind. If the pattern matcher were less broken you'd also have to worry about inferring the kind of a type variable, but you're probably safe there since it will never work.

`def f[CC[X], M[_] <: CC[_]]`

What is the kind of M? What is the relationship between M's type argument and CC's?

For more reasons you must get clear of existentials, see [SI-3713](https://issues.scala-lang.org/browse/SI-3713) and the not really related but I can't resist [SI-7730](https://issues.scala-lang.org/browse/SI-7730) because I still have a backlog of astonishment and most of my tickets have never been seen by a human. Just drink it in.

```
class Foo {
  type T <: String
  // Inferred return type is... Foo.this.T
  def f(x: Any) = x match { case x: (T forSome { type T }) => x }
}
```

Also see [SI-8039](https://issues.scala-lang.org/browse/SI-8039) for other things to improve upon, or at least to not make any worse than they already are (I'd say it's difficult to believe they could get worse, but it isn't.)

**Comment by non** [comment]
When you say "unmake the mistake of confusing existential type arguments with unnamed higher order type parameters" what course of action are you encouraging? I agree with the sentiment but I'm not sure what concrete action would make sense, given compatibility constraints?

Unless you are trying to teach a _War Games_-esque lesson about the only winning move being not to play.

**Comment by paulp** [comment]
@non I don't pretend to understand the compatibility constraints. The issues which have been opened thus far would in aggregate render this version of the language hopelessly incompatible by any measure I can think of. Somewhere it was said that you aren't striving for source compatibility, only binary, which is good because you have zero chance of source compatibility, but binary compatibility can mean a dozen different things so even that doesn't tell me much. Anyway, since I don't understand the constraints you have to interpret anything I say through the lens of "here's a good idea if you can use it."

Since this ticket is talking about a syntax for partial type application, my comments are aimed at what it would take to introduce such a syntax without adding a force multiplier to the complexity of type signatures. The existential underscore confusion is without any doubt a mistake, but the damage done has been limited by the absence of any partial application syntax. If you add it without unmaking that mistake you will reap a whirlwind.

And yes the subtext is always the wargames subtext, but I try only to chime in with concrete and relevant considerations.

**Comment by paulp** [comment]
And since I didn't really answer your question: you have to use a different identifier for existentials, but it's existentials which should get ?. If existentials kept the _ and partial application took ? it would be exactly backward from what anyone would expect from all available clues.

**Comment by non** [comment]
Yeah you're 100% right about that. Kind Projector does the only thing it can, but we should not repeat that here.

**Comment by non** [comment]
I think our only hope of "changing" any of this is to use a language flag, which would have the effect of majorly-breaking source compatibility when it was enabled.

For these reasons I think it may make sense to leave the placeholder syntax out altogether for now.

Given that people are actually writing type projections "in the wild" I think giving them a better syntax for that (even something moderately compromised like `[A] => ...`) functions as a form of harm reduction.

**Comment by paulp** [comment]
For a non-placeholder syntax I strongly urge you to keep looking for something which doesn't mandate the introduction of four characters of punctuation.

For instance, `(A)Either[A, Int]`, or if you can't resist more square brackets, option 1 from above, `[A]Either[A, Int]`.

**Comment by mpilquist** [comment]
For what it's worth, `[A]Either[A, Int]` is consistent with the way scalac currently renders type lambdas:

``` scala
scala> Functor[({type l[a] = String \/ a})#l]
res0: scalaz.Functor[[a]scalaz.\/[String,a]] = scalaz.DisjunctionInstances1$$anon$1@3fe9fd4a
```

**Comment by paulp** [comment]
@mpilquist It's also how they're rendered internally:

```
  type M[A] = Functor[({type l[a] = A \/ a})#l]
  // -Xprint:typer
  // type M[A] = Functor[[a]\/[A,a]]
```

...admittedly the usual rule is that if scalac does it some way internally, it's the wrong way. But there are exceptions and this is a likely one.

**Comment by paulp** [comment]
@mpilquist Oh, that's what you said. Somehow I read that as "how scalaz renders type lambdas."

**Comment by gclaramunt** [comment]
I still have to think about the implications, but multiple parameter list for types can be handy:
T[A][B] = ....
then
T[A] = ({type λ[b] = T[A, b]})#λ 
and
T =  ({type λ[a] = T[a]})#λ = ({type λ[a] = ({type λ[b] = T[a, b]})#λ})#λ

**Comment by paulp** [comment]
@gclaramunt Way past handy. Also, varargs type parameters. If we had these things we wouldn't be able to imagine giving them up. I wouldn't really bet on seeing those things in this project, but I'm optimistic about seeing them in mine.

**Comment by paulp** [comment]
Also, named type arguments. Also, the "infer me" type argument.
`f[?, foo.type, ?, ?](foo)`
as compared to when you have to supply a single type argument let's now go figure out what all those other damn types were inferred to be, and there is no limit to how difficult they might be to write down.

**Comment by paulp** [comment]
Of course those combine nicely when we infer to infer the missing ones.
`f[Coll = foo.type](foo)`

**Comment by mandubian** [comment]
What is the status of this discussion?
Simplifying the type lambda syntax is mandatory IMHO to make scala simply better (and then tackle the type lambda inconsistencies in scalac)

As some people said, `[A] Either[A, Int]` corresponds to what scalac shows but I can accept other syntax.

**Comment by aryairani** [comment]
Named / "infer me" type arguments would have made many past days less frustrating for me.

@mandubian my read on this thread is that `[a]Either[a,Int]` is the current leader, or `Either[_,Int]` for simple definitions and `[a,b]LensFamily[a,a,b,b]` for more complex ones.  `[a,b](Int => a => b)` ?

**Comment by puffnfresh** [comment]
Closed via #70.
